File: packages/php/src/RateLimiter.php
<?php

declare(strict_types=1);

namespace Uncaught;

/**
 * Sliding-window rate limiter.
 *
 * Enforces:
 *  - A global maximum of events per 60-second window.
 *  - A per-fingerprint maximum of events per 60-second window.
 */
class RateLimiter
{
    /** Window duration in seconds. */
    private const WINDOW_SECONDS = 60;

    private int $globalMax;
    private int $perFingerprintMax;

    /** @var float[] Global timestamps (microtime). */
    private array $globalTimestamps = [];

    /** @var array<string, float[]> Per-fingerprint timestamp arrays. */
    private array $buckets = [];

    public function __construct(int $globalMax = 30, int $perFingerprintMax = 5)
    {
        $this->globalMax = $globalMax;
        $this->perFingerprintMax = $perFingerprintMax;
    }

    /**
     * Returns true if the event is allowed through, false if it should be dropped.
     */
    public function shouldAllow(string $fingerprint): bool
    {
        $now = microtime(true);
        $cutoff = $now - self::WINDOW_SECONDS;

        // Prune and check global limit
        $this->globalTimestamps = array_values(
            array_filter($this->globalTimestamps, fn(float $t) => $t > $cutoff)
        );

        if (count($this->globalTimestamps) >= $this->globalMax) {
            return false;
        }

        // Prune and check per-fingerprint limit
        if (!isset($this->buckets[$fingerprint])) {
            $this->buckets[$fingerprint] = [];
        }

        $this->buckets[$fingerprint] = array_values(
            array_filter($this->buckets[$fingerprint], fn(float $t) => $t > $cutoff)
        );

        if (count($this->buckets[$fingerprint]) >= $this->perFingerprintMax) {
            return false;
        }

        // Record this event
        $this->globalTimestamps[] = $now;
        $this->buckets[$fingerprint][] = $now;

        // Periodic cleanup: remove empty buckets
        if (count($this->buckets) > 1000) {
            $this->buckets = array_filter(
                $this->buckets,
                fn(array $ts) => count($ts) > 0
            );
        }

        return true;
    }
}

File: packages/rust/src/rate_limiter.rs
// ---------------------------------------------------------------------------
// uncaught — sliding-window rate limiter (thread-safe)
// ---------------------------------------------------------------------------

use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{Duration, Instant};

/// Window duration (60 seconds).
const WINDOW_DURATION: Duration = Duration::from_secs(60);

/// Sliding-window rate limiter that enforces:
///  - A global maximum of events per 60-second window.
///  - A per-fingerprint maximum of events per 60-second window.
pub struct RateLimiter {
    global_max: usize,
    per_fingerprint_max: usize,
    state: Mutex<RateLimiterState>,
}

struct RateLimiterState {
    global_timestamps: Vec<Instant>,
    buckets: HashMap<String, Vec<Instant>>,
}

impl RateLimiter {
    /// Create a new rate limiter.
    ///
    /// - `global_max`: Max events across all fingerprints per window.
    /// - `per_fingerprint_max`: Max events for a single fingerprint per window.
    pub fn new(global_max: usize, per_fingerprint_max: usize) -> Self {
        Self {
            global_max,
            per_fingerprint_max,
            state: Mutex::new(RateLimiterState {
                global_timestamps: Vec::new(),
                buckets: HashMap::new(),
            }),
        }
    }

    /// Returns `true` if the event identified by `fingerprint` is allowed
    /// through, `false` if it should be dropped.
    pub fn should_allow(&self, fingerprint: &str) -> bool {
        let mut state = match self.state.lock() {
            Ok(s) => s,
            Err(_) => return false,
        };

        let now = Instant::now();
        let cutoff = now.checked_sub(WINDOW_DURATION).unwrap_or(now);

        // Prune and check global limit
        state.global_timestamps.retain(|&t| t > cutoff);
        if state.global_timestamps.len() >= self.global_max {
            return false;
        }

        // Prune and check per-fingerprint limit
        {
            let fp_timestamps = state
                .buckets
                .entry(fingerprint.to_string())
                .or_insert_with(Vec::new);
            fp_timestamps.retain(|&t| t > cutoff);

            if fp_timestamps.len() >= self.per_fingerprint_max {
                return false;
            }
        }

        // Record this event
        state.global_timestamps.push(now);
        state.buckets.get_mut(fingerprint).unwrap().push(now);

        // Periodic cleanup: remove empty buckets
        if state.buckets.len() > 1000 {
            state.buckets.retain(|_, ts| !ts.is_empty());
        }

        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_allows_within_limit() {
        let limiter = RateLimiter::new(10, 3);
        assert!(limiter.should_allow("fp1"));
        assert!(limiter.should_allow("fp1"));
        assert!(limiter.should_allow("fp1"));
    }

    #[test]
    fn test_blocks_per_fingerprint() {
        let limiter = RateLimiter::new(10, 2);
        assert!(limiter.should_allow("fp1"));
        assert!(limiter.should_allow("fp1"));
        // Third should be blocked
        assert!(!limiter.should_allow("fp1"));
        // Different fingerprint should still be allowed
        assert!(limiter.should_allow("fp2"));
    }

    #[test]
    fn test_blocks_global() {
        let limiter = RateLimiter::new(3, 10);
        assert!(limiter.should_allow("fp1"));
        assert!(limiter.should_allow("fp2"));
        assert!(limiter.should_allow("fp3"));
        // Fourth should be blocked globally
        assert!(!limiter.should_allow("fp4"));
    }
}
